src/third/redux-persist/createPersistor.js

```diff
@@ -73,20 +78,22 @@ export default function createPersistor (store, config) {
    */
   async function writeOnce(state) {
     // Identify what keys we need to write.
-    const updatedKeys = [];
+    // This includes anything already in outstandingKeys, because we don't
+    // know what value was last successfully stored for those.
     for (const key of Object.keys(state)) {
       if (!passWhitelistBlacklist(key)) {
         continue;
       }
       if (state[key] === lastWrittenState[key]) {
         continue;
       }
-      updatedKeys.push(key);
+      outstandingKeys.push(key);
     }
+    lastWrittenState = state
 
     // Serialize those keys' subtrees, with yields after each one.
     const writes = []
-    for (const key of updatedKeys) {
+    for (const key of outstandingKeys) {
```

Comment:
> So it's now possible for `storage.multiSet` to be called with duplicated keys, right? A previous write to 'foo' failed (to report success anyway), and this time 'foo' was also updated and so we need to write the new value.

Hmmm, good catch, thanks!

The fix is that `outstandingKeys` should be a `Set` rather than an array.

I actually drafted this initially with a `Set`, following what I'd said at https://github.com/zulip/zulip-mobile/pull/4694#discussion_r691727392 . Then I thought, "hmm, why does this need to be a set? We never make a `.has` call on it. It can just be an array. I must have been thinking of some wrinkle that we didn't end up needing to have." So I changed it to an array. But that reasoning was wrong, because even if you never call `has`, `add` on a set has one important difference from `push` on an array! Namely, it drops duplicates.